wcfsetup/install/files/lib/util/UserAgent.class.php

```diff
@@ -0,0 +1,407 @@
+		if (preg_match('~(?:crios|crmo)/([\d\.]+)|chrome/([\d\.]+).*mobile~i', $this->userAgent, $match)) {
+			$this->browser = 'Chrome Mobile';
+			$this->browserVersion = (isset($match[2]) ? $match[2] : $match[1]);
+			return;
+		}
+		
+		// kindle
+		if (preg_match('~kindle/([\d\.]+)~i', $this->userAgent, $match)) {
+			$this->browser = 'Kindle';
+			$this->browserVersion = $match[1];
+			return;
+		}
+		
+		// silk
+		if (preg_match('~silk/([\d\.]+)~i', $this->userAgent, $match)) {
+			$this->browser = 'Silk';
+			$this->browserVersion = $match[1];
+			return;
+		}
+		
+		// android browser
+		if (preg_match('~Android ([\d\.]+).*AppleWebKit~i', $this->userAgent, $match)) {
+			$this->browser = 'Android Browser';
+			$this->browserVersion = $match[1];
+			return;
+		}
+		
+		// safari mobile
+		if (preg_match('~([\d\.]+) Mobile/\w+ safari~i', $this->userAgent, $match)) {
+			$this->browser = 'Safari Mobile';
+			$this->browserVersion = $match[1];
+			return;
+		}
+		
+		// chrome
+		if (preg_match('~(?:chromium|chrome)/([\d\.]+)~i', $this->userAgent, $match)) {
+			$this->browser = 'Chrome';
+			$this->browserVersion = $match[1];
+			return;
+		}
+		
+		// safari
+		if (preg_match('~([\d\.]+) safari~i', $this->userAgent, $match)) {
+			$this->browser = 'Safari';
+			$this->browserVersion = $match[1];
+			return;
+		}
+	}
+	
+	/**
+	 * Detects the OS on the basis of the user agent.
+	 */
+	protected function detectOs(): void {
+		// iOS 
+		if (preg_match('/iphone/i', $this->userAgent)) {
+			$this->os = "iOS";
+			return;
+		}
+		
+		// iOS 
+		if (preg_match('/cfnetwork/i', $this->userAgent)) {
+			$this->os = "iOS";
+			return;
+		}
+		
+		// Windows 
+		if (preg_match('/windows/i', $this->userAgent)) {
+			$this->os = "Windows";
+			return;
+		}
+		
+		// FreeBSD 
+		if (preg_match('/freebsd/i', $this->userAgent)) {
+			$this->os = "FreeBSD";
+			return;
+		}
+		
+		// netBSD 
+		if (preg_match('/netbsd/i', $this->userAgent)) {
+			$this->os = "NetBSD";
+			return;
+		}
+		
+		// openBSD 
+		if (preg_match('/openbsd/i', $this->userAgent)) {
+			$this->os = "OpenBSD";
+			return;
+		}
+		
+		// Android 
+		if (preg_match('/android/i', $this->userAgent)) {
+			$this->os = "Android";
+			return;
+		}
+		
+		// Linux 
+		if (preg_match('/linux/i', $this->userAgent)) {
+			$this->os = "Linux";
+			return;
+		}
+		
+		// iPad 
+		if (preg_match('/ipad/i', $this->userAgent)) {
+			$this->os = "iPad";
+			return;
+		}
+		
+		// webOS 
+		if (preg_match('/web0s/i', $this->userAgent)) {
+			$this->os = "webOS";
+			return;
+		}
+		
+		// CrOS 
+		if (preg_match('/cros/i', $this->userAgent)) {
+			$this->os = "Chrome OS";
+			return;
+		}
+		
+		// macOS 
+		if (preg_match('/mac/i', $this->userAgent)) {
+			$this->os = "macOS";
+			return;
+		}
+	}
+	
+	/**
+	 * Returns the browser based on the user agent or null, if no browser can be determined.
+	 */
+	public function getBrowser(): ?string {
+		if ($this->browser === null) $this->detectBrowser();
```

Comment:
If no browser could be detected the detection will run over and over again. It might make sense to move the detection into the constructor. Usually developers create `UserAgent` objects to later pull out the information anyway. There is no need make the detection lazy, it need to happen anyway.